### PR TITLE
Feature/peripheral documentation

### PR DESCRIPTION
**Pull request by Kacper20**
Following changes has happened:
- Work on Example app - it's almost finished, needs a little bit polishing & writing to characteristic feature.
- Eliminated race conditions in several calls that could happen.
- Created documentation for `Peripheral` class.

### CODE REVIEW

File: Example/Example/CharacteristicsController.swift [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    override func viewWillAppear(animated: Bool) {
+        super.viewWillAppear(animated)
+        getCharacteristicsForService(service)
+    }
+
+    private func getCharacteristicsForService(service: Service) {
+        service.discoverCharacteristics(nil)
+            .subscribeNext { characteristics in
```
Comment (literator) [inline Example/Example/CharacteristicsController.swift]: `[weak self]` as a good practice here?

File: Example/Example/CharacteristicsController.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    private var monitorValueUpdateDisposable: Disposable?
+    private func setNotificationsState(enabled enabled: Bool, characteristic: Characteristic) {
+        monitorValueUpdateDisposable?.dispose()
+
+        monitorValueUpdateDisposable = characteristic.setNotifyValue(enabled)
+            .doOnNext { self.refreshCharacteristic($0) }
+            .flatMap { $0.monitorValueUpdate() }
+            .subscribeNext {
+                self.refreshCharacteristic($0)
```
Comment (literator) [inline Example/Example/CharacteristicsController.swift]: Consider `[weak self]`.

Btw. you can call this like that:
```swift
.subscribeNext(refreshCharacteristic)
```

File: Example/Example/CharacteristicsController.swift [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        if characteristic.properties == .Read {
+            let readValueNotificationAction = UIAlertAction(title: "Trigger value read", style: .Default) { _ in
+                self.triggerValueReadForCharacteristic(characteristic)
+            }
+            actionSheet.addAction(readValueNotificationAction)
+        }
+
+        if characteristic.properties == .Write {
+            let writeValueAction = UIAlertAction(title: "Write value", style: .Default) { _ in
+
```
Comment (literator) [inline Example/Example/CharacteristicsController.swift]: Empty? Maybe put a comment here or smt?

File: Example/Example/CharacteristicsController.swift [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        valueWriteController.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: nil))
+        valueWriteController.addAction(UIAlertAction(title: "Write", style: .Default) { _ in
+            print("")
+        })
+    }
+
+    private func triggerValueReadForCharacteristic(characteristic: Characteristic) {
+        characteristic.readValue()
+            .subscribeNext {
+                self.refreshCharacteristic($0)
```
Comment (literator) [inline Example/Example/CharacteristicsController.swift]: `[weak self]`
or
```swift
.subscribeNext(refreshCharacteristic) // even more elegant
```

File: Example/Example/CharacteristicsController.swift [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    func updateWithCharacteristic(characteristic: Characteristic) {
+        self.UUIDLabel.text = characteristic.uuid.UUIDString
+        self.isNotifyingLabel.text = characteristic.isNotifying ? "true" : "false"
+        self.valueLabel.text = characteristic.value?.hexadecimalString ?? "Empty"
+    }
+}
+
+
+
+
```
Comment (literator) [inline Example/Example/CharacteristicsController.swift]: LOTS OF SPACES

File: Example/Example/CharacteristicsTableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+//  CharacteristicsTableViewController.swift
+//  Example
+//
+//  Created by Kacper Harasim on 14.04.2016.
+//  Copyright © 2016 Polidea. All rights reserved.
+//
+
+import Foundation
+
+c
```
Comment (literator) [inline Example/Example/CharacteristicsTableViewController.swift]: c?

File: Example/Example/Main.storyboard [hunk 7]
```diff
@@ -91,25 +92,151 @@
...
             </objects>
-            <point key="canvasLocation" x="1037" y="84"/>
+            <point key="canvasLocation" x="1085" y="71"/>
+        </scene>
+        <!--Characteristics Controller-->
+        <scene sceneID="wzW-Fd-4Tn">
+            <objects>
+                <viewController id="a6s-hS-P2J" customClass="CharacteristicsController" customModule="Example" customModuleProvider="target" sceneMemberID="viewController">
+                    <layoutGuides>
+                        <viewControllerLayoutGuide type="top" id="nFV-LN-dsq"/>
```
Comment (literator) [inline Example/Example/Main.storyboard]: Reviewing storyboard.

84c5461a3eb849468173455fcb621c642a932c50d66e9c14452e562c974e0569

File: README.md [hunk 8]
```diff
@@ -288,9 +294,15 @@ Library supports **complex** Bluetooth error handling functionalities. Errors fr
...
+
+
 ## Contributing
 If you would like to contribute code you can do so through GitHub by forking the repository and sending a pull request.
-To keep code in order, we advice you to use SwiftLint. In repository, we provide configured `.swiftlint.yml` file, that matches our criteria of clean and "Swifty" code.  
+To keep code in order, we advice you to use SwiftLint. In repository, we provide configured `.swiftlint.yml` file, that matches our criteria of clean and "Swifty" code.
+
+### Contributors, thanks!
+
+Maciek Oczko (maciek.oczko@polidea.com)
```
Comment (literator) [inline README.md]: thx! :)

File: RxBluetoothKit/Characteristic.swift [hunk 9]
```diff
@@ -65,66 +80,84 @@ public class Characteristic {
...
 
-    //MARK: Writing characteristic values
+    //MARK: Descriptors
 
     /**
-     Function which monitors writes to this characteristic after subscription to
-     returned observable.
+     Function that triggers descriptors discovery for characteristic.
+     - returns: Observable that emits `Next` with array of `Descriptor` instances, once they're discovered.
+     Immediately after that `.Complete` is emitted.
```
Comment (literator) [inline RxBluetoothKit/Characteristic.swift]: `Complete` without a `.`? `Next` is without.

File: RxBluetoothKit/Characteristic.swift [hunk 10]
```diff
@@ -65,66 +80,84 @@ public class Characteristic {
...
+     */
+    public func discoverDescriptors() -> Observable<[Descriptor]> {
+        return self.service.peripheral.discoverDescriptorsForCharacteristic(self)
+    }
 
-     - returns: Observable which emits this characteristic if write operation was
-                performed on it.
+    /**
+     Function that allow to monitor writes that happened for characteristic.
+     - Returns: Observable that emits `.Next` with `Characteristic` instance every time when write has happened.
```
Comment (literator) [inline RxBluetoothKit/Characteristic.swift]: And here you have `.Next` not `Next`?

File: RxBluetoothKit/Characteristic.swift [hunk 11]
```diff
@@ -65,66 +80,84 @@ public class Characteristic {
...
+     Function that triggers write of data to characteristic. Write is called after subscribtion to `Observable` is made.
+     Behavior of this function strongly depends on [CBCharacteristicWriteType](https://developer.apple.com/library/ios/documentation/CoreBluetooth/Reference/CBPeripheral_Class/#//apple_ref/swift/enum/c:@E@CBCharacteristicWriteType), so be sure to check this out before usage of the method.
+     - parameter forCharacteristic: `Descriptor` instance to write value to.
+     - parameter type: Type of write operation. Possible values: `.WithResponse`, `.WithoutResponse`
+     - returns: Observable that emition depends on `CBCharacteristicWriteType` passed to the function call.
+     Behavior is following:
+
+     - `WithResponse` -  Observable emits `.Next` with `Characteristic` instance write was confirmed without any errors.
+     Immediately after that `Complete` is called. If any problem has happened, errors are emitted.
+     - `WithoutResponse` - Observable emits `.Next` with `Characteristic` instance once write was called.
```
Comment (literator) [inline RxBluetoothKit/Characteristic.swift]: So you put `.` prefix - ok, check above, you forgot somewhere to add id.

File: RxBluetoothKit/DeviceIdentifiers.swift [hunk 12]
```diff
@@ -22,17 +22,71 @@
...
 
-protocol ServiceIdentifier {
-    var UUID: CBUUID { get }
-}
+/**
+ ServiceIdentifier protocol specifies how information about services of device should look like.
+ Sample implementation could look like:
 
-protocol DescriptorIdentifier {
+ ```
```
Comment (literator) [inline RxBluetoothKit/DeviceIdentifiers.swift]: swift

Comment (literator) [inline RxBluetoothKit/DeviceIdentifiers.swift]: swift?

File: RxBluetoothKit/DeviceIdentifiers.swift [hunk 13]
```diff
@@ -22,17 +22,71 @@
...
     var UUID: CBUUID { get }
-    var characteristic: CharacteristicIdentifier { get }
 }
 
-protocol CharacteristicIdentifier {
+/**
+ Characteristic protocol specifies how information about characteristics of device should look like.
+ Sample implementation could look like:
+
+ ```
```
Comment (literator) [inline RxBluetoothKit/DeviceIdentifiers.swift]: swift? not sure - it's in source

File: RxBluetoothKit/Peripheral.swift [hunk 14]
```diff
@@ -105,18 +124,21 @@ public class Peripheral {
         })
         .take(1)
 
-        return Observable.deferred {
-            self.peripheral.discoverServices(identifiers)
-            return self.ensureValidPeripheralState(observable)
+        return Observable.create { observer in
+            self.ensureValidPeripheralState(observable).subscribe(observer)
```
Comment (literator) [inline RxBluetoothKit/Peripheral.swift]: Is here a disposable handled somehow?

File: RxBluetoothKit/Peripheral.swift [hunk 15]
```diff
@@ -162,16 +186,18 @@ public class Peripheral {
         }
         .take(1)
 
-        return Observable.deferred {
+        return Observable.create { observer in
+            self.ensureValidPeripheralState(observable).subscribe(observer)
```
Comment (literator) [inline RxBluetoothKit/Peripheral.swift]: The same question about disposable here?

File: RxBluetoothKit/Peripheral.swift [hunk 16]
```diff
@@ -186,32 +212,42 @@ public class Peripheral {
...
             switch type {
             case .WithoutResponse:
-                return self.ensureValidPeripheralState(Observable.just(characteristic))
+                self.peripheral.writeValue(data, forCharacteristic: characteristic.characteristic, type: type)
+                self.ensureValidPeripheralState(Observable.just(characteristic)).subscribe(observer)
             case .WithResponse:
-                return self.ensureValidPeripheralState(self.monitorWriteForCharacteristic(characteristic).take(1))
+                self.peripheral.writeValue(data, forCharacteristic: characteristic.characteristic, type: type)
+                self.ensureValidPeripheralState(self.monitorWriteForCharacteristic(characteristic).take(1))
+                    .subscribe(observer)
```
Comment (literator) [inline RxBluetoothKit/Peripheral.swift]: And here.

**Comment by literator** [comment]
Mostly good. Docs are awesome. Check out comments. The biggest issue is that you don't handle disposables properly (I think) in a few places. Take a look at it.

**Comment by Kacper20** [comment]
Hi, many thanks for review, great job :-) These comments were really helpful - I handled disposable in a way which we were talking about.

**Pull request merged**